fairscale/nn/misc/auto_wrap.py

```diff
@@ -0,0 +1,197 @@
+            self.l1 = auto_wrap(TransformerBlock(), min_num_params=1e8)
+
+    Args:
+        module (nn.Module): module to wrap (if in :func:`enable_wrap` context)
+        cls (Callable): class wrapper to wrap the model with if in context
+            (default: :class:`FullyShardedDataParallel`)
+        min_num_params (int, Optional): min number of parameters for a child
+            Module to be wrapped
+        activation_checkpoint (bool): use activation checkpointing wrapper
+            (default: False)
+    """
+    if ConfigAutoWrap.in_autowrap_context:
+        wrapped_module, remainder = ConfigAutoWrap.recursive_wrap(
+            module, cls=cls, activation_checkpoint=activation_checkpoint, min_num_params=min_num_params, **kwargs
+        )
+        return wrapped_module
+    return module
+
+
+class ConfigAutoWrap:
+    """
+    Helper class to wrap modules based on default config args via a context manager.
+    See :func:`enable_wrap` for more information.
+    """
+
+    module_blocklist: List = []
+    in_autowrap_context: bool = False
+    kwargs: Dict[str, Any] = {}
+
+    def __init__(self, module_blocklist: Optional[List] = None, **kwargs: Dict[str, Any]):
+        if module_blocklist:
+            self.module_blocklist += module_blocklist
+        self.kwargs = kwargs
+
+    @staticmethod
+    def enable_autowrap_context(kwargs: Any) -> None:
+        ConfigAutoWrap.in_autowrap_context = True
+        ConfigAutoWrap.kwargs = kwargs
+        ConfigAutoWrap.module_blocklist += FSDP_MODULE_BLOCKLIST
+
+    @staticmethod
+    def disable_autowrap_context() -> None:
+        ConfigAutoWrap.in_autowrap_context = False
+        ConfigAutoWrap.kwargs = {}
+        ConfigAutoWrap.module_blocklist = []
+
+    def __enter__(self) -> None:
+        self.enable_autowrap_context(self.kwargs)
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self.disable_autowrap_context()
+
+    @staticmethod
+    def recursive_wrap(module: nn.Module, min_num_params: Union[int, float], **kwargs: Any) -> Tuple[nn.Module, int]:
+        """
+        Automatically wrap child modules of *module* that meet the given
+        criteria with :func:`auto_wrap`.
+
+        Args:
+            module (nn.Module): module to recursively wrap
+            min_num_params (int): min number of parameters for a child Module to
+                be wrapped
+        """
+        if isinstance(module, tuple(ConfigAutoWrap.module_blocklist)):
+            # If the module has been blocklisted from wrapping, we return
+            return module, 0
+
+        num_params = sum([p.numel() for p in module.parameters()])
+
+        if len(list(module.named_children())) == 0:
+            # If the module has no children, no need to recurse, wrap it if needed
+            if num_params >= min_num_params and not isinstance(module, tuple(FSDP_MODULE_EXCLUDE_WRAP)):
+                return wrap(module, **kwargs), num_params
+            return module, 0
```

Comment:
I don't know if this is needed. when there is 0 child, we take care of it on line 183 by not looping over the children list, right?

Reply:
Think you're right, great observation!